### PR TITLE
More consistent parsing of boolean inputs (Breaking change)

### PR DESCRIPTION
**Pull request by bendudson**
TL;DR: Only "true" and "false", "True" and "False", 1 and 0 are allowed for booleans. This enables boolean options expressions like:
```
thing = true
another = false
third = thing & another
```
but breaks inputs using "yes", and other strings that are not "true" or "false", "True", "False", "1" or "0" for boolean settings.

When converting a string to a BoutReal, Options uses an expression parser that can use variables defined in the input options. The same parser is used when converting to an integer. When converting to a bool, however, Options was doing a simple case-insensitive match to "true", "false", "yes", "no" and a few similar strings.

This PR changes this behaviour, treating bools like a special case of an integer: The same parser is used, but the result must be either 1 or 0. This enables boolean quantities to behave consistently with int and BoutReal types.

We now define "true" to be equal to 1, and "false" to be 0, so can be used in expressions e.g `true * 2 -> 2`, and `false * 3 -> 0`. No other variations so e.g. "yes", "no", "True" are undefined.

We define operators `|` and `&` for logical `or`, and logical `and`; and `!` for logical `not`. Due to the simplicity of the parser binary operators must be single characters.  These both check that their left and right arguments are bools (i.e. either 1 or 0). Also adding operators `<` and `>` for comparing values, evaluating to 1 (true) or 0 (false).

Previously x,y,z and t were initialized to 0, but for booleans this would now have quite bad consequences: Inputs that used "y" to mean "true" would now evaluate to `0` which would be interpreted as `false`. With this change we get an exception ("y" is undefined)
rather than incorrect behaviour.

By removing these defaults, scalar quantities (e.g. integers, booleans) are evaluated without defining x,y,z or t. Trying to convert an expression like "x + y" to an integer or bool would previously evaluate to 0, but will now throw a BoutException.

### CODE REVIEW

**Comment by dschwoerer** [comment]
I guess this also resolves https://github.com/boutproject/BOUT-dev/issues/1701

**Review by dschwoerer — DISMISSED** [review]
I feel slightly uneasy about such a breaking change in a 5.2.0 release.

However:
 * The change should result in errors, not in subtle changes
 * The change might surface bugs
 * In general at least, it should be easy to update to the new syntax

Should this also be documented?

**Comment by ZedThree** [comment]
LGTM! I think it's really nice to tighten up the input file format, as well as fixing other issues and adding new features!

It would be good to have this change documented somewhere -- there's a breaking changes section in the changelog for example.

Is there some way we could deprecate `yes/no/True/False` etc? Maybe we could add generators for them that print warnings if they get used?
